datastore/emulator.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datastore
+
+import (
+	"bytes"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"net/http"
+	"os"
+	"os/exec"
+	"syscall"
+)
+
+// Datastore emulator for use in tests
+//
+// Background: this was devised due to aetest package no longer being available for 1.12+ runtimes as dev_appserver.py
+// cannot run runtimes past go111
+//
+//   Usage:
+//
+//   DSStop, DSReset := Emulator(8081)
+//   defer DSStop()
+//
+func Emulator(port int) (stop func(), reset func()) {
```

Comment:
Do we need to start/stop this from Go tests, and re-start it for every test? If not, perhaps it would be simpler to started the emulator externally, before running `go test` (using a Makefile or a bash script)

Reply:
Unfortunately current tests require a separate env as some of them pollute the namespace which can cause issues for other tests. So it's either this or setting up some sort of cleanup routine in `testing.M` WDYT?

Reply:
I think refactoring tests to allow them to run concurrently (or adding test cleanup) will be simpler than managing and restarting an emulator subprocess from Go.

Reply:
Chatted about this offline - agreed to keep the emulator but it's coming out or getting immediately fixed if it gets flaky.